### PR TITLE
Fix run now errors and interrupts

### LINKED ISSUE #191: Error handling is no longer correct
Multiple tests in `test-run_now.R` are failing; we didn't catch this because they're disabled on ci.

In `test-run_now.R`, there are tests that test what happens if a callback:

1. throws an error while `run_now` is called
2. is interrupted while `run_now` is called

Tests in the second category apparently don't work on CRAN or CI environments. But tests in the first should be fine, and they've started failing since this commit in Rcpp (between 1.0.9 and 1.0.10, in 2022):

https://github.com/RcppCore/Rcpp/commit/d389a8a9540690fdc1839507abe95fdfdee8acd7

The problem is that our `invoke_wrapped()` logic assumes that Rcpp's unwind protect feature is not in use, and the default became to use it. With unwind protect, the right thing to do is _almost_ to just call `invoke()` directly and not use `invoke_wrapped()` at all. I say almost, because this works great for errors but not for interrupts--the interrupts somehow aren't caught by `tryCatch(..., interrupt = ...)` so the tests abruptly end.

### PR DESCRIPTION
**Pull request by jcheng5**
Fixes #191

### CODE REVIEW

**Comment by jcheng5** [comment]
Re: remaining failures for R 4.1 and 3.6 on Windows: On these builds, having Rcpp code throw an exception that's caught by the calling R code, seems to be escaping the try/catch somehow. What's strangest is that this happens _only for 32-bit Windows_.

```r
Rcpp::cppFunction(
  code = '
      void cpp_error() {
        throw std::runtime_error("This is a C++ exception");
      }
    '
)

tryCatch(cpp_error(), error = function(e) message("got here"))
```

**Comment by jcheng5** [comment]
Minimal repro looks like the below. It actually _works_ when called from RStudio, but fails when called from R at a terminal or if you put it in a file and call `R -f test.R`.

```r
library(later)

Rcpp::cppFunction(
  depends = "later",
  includes = '
      #include <later_api.h>

      void oof(void* data) {
        throw std::runtime_error("This is a C++ exception.");
      }
    ',
  code = '
      void cpp_error(int value) {
        later::later(oof, 0, 0);
      }
    '
)

tryCatch(
  {cpp_error(1); later::run_now()},
  error = function(e) {message("got here")}
)
```

**Comment by shikokuchuo** [comment]
I notice you're not using `rlang::interrupt()`, so just in case this behaviour is relevant:

``` r
mirai::mirai(rlang::interrupt())[]
#> 'miraiInterrupt' chr ""

mirai::mirai(tools::pskill(Sys.getpid(), tools::SIGINT))[]
#> [1] TRUE
```

<sup>Created on 2024-10-29 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

FYI mirai handles errors entirely on the R side with a combination of with `withRestarts()` and `withCallingHandlers()` here: https://github.com/shikokuchuo/mirai/blob/main/R/daemon.R#L204.

**Comment by shikokuchuo** [comment]
Just to clarify my comment above, and as I've had the chance to test this, if I change the following line https://github.com/r-lib/later/blob/e3239dc35d54588ca9c9f69f9065ff3d44d1c620/tests/testthat/test-run_now.R#L179 to:

```r
later(function() { rlang::interrupt() }) 
```
then those `run_now()` tests pass. Raising SIGINT through `tools::pskill()` was not creating an interrupt condition that was being caught by the tryCatch.

Btw. If I test this PR, it exits early doesn't complete and I'm on R4.4.2 on Ubuntu 22, so it's a bit of a mystery how it passes on CI...

**Comment by jcheng5** [comment]
@shikokuchuo Thank you for the suggestion to use `rlang::interrupt()`. The `pskill` approach was definitely not doing the right thing on Windows, neither was sending Ctrl+C the way I was doing it.

After I removed those, I was able to remove almost all of the testthat skips. The only problem that still remains is that on Windows i386 only, native `later::later` callbacks that raise C++ exceptions that are raised via `later::run_now` blow through my `tryCatch` calls. (The reason why this only breaks on the R 4.1 and 3.6 CI runs is that R 4.2 removed i386!) If we have to ship with testthat skipping those tests for Windows i386, I can probably live with that.

File: src/later.cpp
Comment (shikokuchuo) [reply src/later.cpp:210]: I attempted to implement this in the latest commit here in my later branch: https://github.com/shikokuchuo/later/commits/uwp/

The strange thing is, it appears to work fine both errors and interrupts when I use this with a mirai / Shiny stack (mirai sends stdFunction callbacks to later). And yet it reliably segfaults during tests / R CMD check.

Comment (shikokuchuo) [reply src/later.cpp:210]: An alternative would be to disable Rcpp unwind protect to revert to the old behaviour (also suggested by Lionel). The safest way is probably by adding the flags to Makevars.in and Makevars.win as I've done in the latest commit here: https://github.com/shikokuchuo/later/commits/disable_uwp/ Note that I've had to skip 2 of the old interrupt tests for clean check results.

Comment (shikokuchuo) [reply src/later.cpp:210]: Making some more progress to help get this over the line:

1. I can confirm that we do not need to `unwindProtect()` the `Rcpp::Function` invocations as these call `unwindProtect()` anyway: https://github.com/RcppCore/Rcpp/blob/6193de530785f1d84117a852cf7e478b1f048b2f/inst/include/Rcpp/Function.h#L82 and https://github.com/RcppCore/Rcpp/blob/6193de530785f1d84117a852cf7e478b1f048b2f/inst/include/Rcpp/api/meat/Rcpp_eval.h#L61-L63.

2. I've modified our `std::function<void(void)>` case to exactly mirror that for `std::function<SEXP(void)>` present in Rcpp's source code at https://github.com/RcppCore/Rcpp/blob/6193de530785f1d84117a852cf7e478b1f048b2f/inst/include/Rcpp/unwindProtect.h#L80. This still segfaults for `R CMD check`, but interestingly does not for `devtools::test()`, although there are 7 test failures (the same failures that are present without my changes on this branch).

My `uwp` branch has been updated: https://github.com/shikokuchuo/later/commits/uwp/.

Comment (jcheng5) [reply src/later.cpp:210]: This is all super helpful. @shikokuchuo other than disabling unwind protect, have you found a usage of `unwindProtect()` that fixes my example here? https://github.com/r-lib/later/pull/192#issuecomment-2438598770

Comment (shikokuchuo) [reply src/later.cpp:210]: @lionel- the comment to your original commit to Rcpp is along the lines of what we discussed today that we should unwind protect arbitrary code invocatons: https://github.com/RcppCore/Rcpp/commit/995f17b7db4b5b8cd2dcc8b700b9587c25ef5d68#diff-91c5b46dc84a94604a4e4d0caed9bf85590a2eddbb12d2e8dc80badf324a9dfbR2

However, this is the output from running `R CMD check --use-valgrind` on my latest `uwp` branch: https://gist.github.com/shikokuchuo/a5e4396b4d389276902f732d69a0790d

Specifically looking at `by 0x12DFFEBF: _later_execCallbacks.cold (RcppExports.cpp:130)`
That RcppExports line 130 is actually the `END_RCPP` macro. Specifically, given the `exception_to_r_condition` call, it is this branch of the `END_RCPP` TryCatch: https://github.com/RcppCore/Rcpp/blob/6193de530785f1d84117a852cf7e478b1f048b2f/inst/include/Rcpp/macros/macros.h#L76. So it seems a `std::exception` is being caught rather than a `Rcpp::LongjumpException`. Perhaps this is where we want to get to, after the long jump is unwound, but then getting the last call information somehow leads to the invalid read.

Valgrind actually passes on *this* branch (rather than my `uwp` branch with the modifications from today).

Comment (shikokuchuo) [reply src/later.cpp:210]: > This is all super helpful, thank you both. @shikokuchuo other than disabling unwind protect, have you found a usage of `unwindProtect()` that fixes my example here? [#192 (comment)](https://github.com/r-lib/later/pull/192#issuecomment-2438598770)

Do you mean for 4.1 and earlier on Windows 32-bit? The example works using your branch here on R4.4 Ubuntu and R4.2 on Windows.

Comment (jcheng5) [reply src/later.cpp:210]: @shikokuchuo Yes, sorry, I did mean for R 4.1 and earlier on Windows 32-bit.

**Comment by jcheng5** [comment]
@shikokuchuo Due to a record-setting windstorm here, I have no access to Windows at the moment; my only Windows PC is a desktop and I have to leave the house with my laptop to have internet. I also will be in meetings for most of today that will require my undivided attention.

Given that you already understand what's happening here much better than I do, and have also talked to @lionel-, can I leave it to you to decide the path forward for this release? As long as it works for all supported platforms including Windows 32-bit. And ideally passing the tests that are currently on this branch (rather than on main, which are using pskill and also skipping all the hard tests on ci and cran)

File: src/later.cpp
Comment (shikokuchuo) [reply src/later.cpp:210]: The tests that are failing are the last batch of C++ tests, where we construct R functions that call Rcpp code, which in turn call later on one of these functions that error in various ways :)

It turns out that this may be too much to wrap in `Rcpp::unwindProtect()` - specifically calling into the Rcpp function may be the issue.

To confirm that C code is behaving as we expect (and confirming the anecdotal evidence that mirai/Shiny seems to be working fine) - I skip the C++ tests, and add one of my own that uses nanonext and abuses an internal function that causes a long jump during an `Rf_eval()` from the later callback invocation. Test results here: https://github.com/shikokuchuo/later/actions/runs/11935403862/job/33267183990. Notably it passes for R4.1 on Windows uniformly for both 64bit and 32bit.

So the remaining to do is to investigate why `Rcpp::unwindProtect()` around an Rcpp function isn't safe.

Comment (jcheng5) [reply src/later.cpp:210]: In my hacking last night I was just barely able to get the C++ test (`throw std::runtime_error("This is a C++ exception.");`) to pass on 32 bit, by putting `BEGIN_CPP`/`END_CPP` inside `oof()`. (Or maybe I had to introduce a wrapper function to get it to compile, but with that general idea).

Anyway, that worked fine, but I replaced `BEGIN_CPP`/`END_CPP` with `unwindProtect`, expecting it to do the same thing, but the latter did not work.

Comment (shikokuchuo) [reply src/later.cpp:210]: Amazing! I do believe we've done it the 3 of us combined.

I had tried putting these macros around the unwindProtect. But of course it's the other way that works. And the reason it does is the `RCPP_END` macro converts all C++ exceptions to R errors, which are then caught by the unwindProtect (as Lionel succinctly put it earlier in the day). Without that, it seems the exceptions were crossing this boundary and causing the segfault.

I've hence wrapped the `std::function` invocation with these macros, before passing to `Rcpp::unwindProtect`. I've also kept all this logic in one place by using a lambda function, which is explicitly supported, in this commit here: https://github.com/shikokuchuo/later/commit/4e9bf3375834688c6927337e01c765665a255b01

This unfortunately still leaves us with the win32 errors, but given the advantages in terms of (i) long-jump protection for C code and (ii) benefits to [avoiding R_TopLevelExec](https://github.com/r-lib/later/pull/192/files#diff-563cfce67d7007e226622761b341eea58237a2c48b84f9f7b42d4248793b85edR132),  I think it's a good trade-off to make. I'll have a word with Lionel in the morning just to see if we can think of why the exceptions don't seem to be propagated up the stack specifically for that platform.

Comment (jcheng5) [reply src/later.cpp:210]: Ohhhh, that makes so much sense when you lay it out like that. Sounds great.

If the Windows x86 problem isn't new, then I don't have a problem shipping this at all. Thanks!

File: src/later.cpp
Comment (jcheng5) [inline src/later.cpp:213]: Do we even need this branch at this point? Is `RCPP_USING_UNWIND_PROTECT` entirely up to us? (I think I misunderstood it as being something configured when Rcpp is installed, but my updated understanding from you both is that this is an option we decide just for ourselves in our own Makevars?)

Comment (shikokuchuo) [reply src/later.cpp:213]: This is defined by the `Rcpp.h` header by default. To turn it off you'd have to define `RCPP_NO_UNWIND_PROTECT` before including the header (or as Makevars). It's true that we don't do this ourselves, but a user could still do it in their user or site Makevars. I think it'd be prudent at least for this release to keep it. That way if there are edge cases that depend on the old behaviour, they can still access this.

File: src/later.cpp
Comment (shikokuchuo) [reply src/later.cpp:210]: Great! I ran this past Lionel and another way to look at it is: "If you're using the unprotected Rcpp callback in an Rcpp stack, all is fine. But by using unwindProtect you introduce an R stack in between. So it should be treated as an R/Rcpp boundary, and the RCPP macros are made for that." He's just great at explanations.

For Win x86 it's never worked AFAICT. I also tried later 1.3.2 and 1.3.0 on an old windows laptop to confirm that it also fails there. Throwing exceptions across module boundaries just doesn't work with Windows SEH (Structured Exception Handling).

On the same topic, Lionel also brought this to my attention: https://stackoverflow.com/questions/13990892/throwing-c-exceptions-outside-static-library If we're catching C++ exceptions thrown by user C++ code, that's also not guaranteed to be safe 100% of the time. But it mostly is if the same compilers etc. are being used. It's also what `later` has always been doing -  using the RCPP macros instead of our own tryCatch block is in no sense worse in this respect.

Whilst it might be worth suggesting in the docs that users can wrap their C++ functions with the RCPP macros for safety (and also ensures it works on Win x86), we've never actually made any requirements on *user* code and I wouldn't want to make that call here. Also looking through this issues, I don't think this has ever come up. In any case, I tested that the current code works fine if users add their own extra layer of `BEGIN_RCPP`/`END_RCPP` in their callback functions.

**Comment by shikokuchuo** [comment]
I'm fine to merge this PR if you're ok with my replies @jcheng5. Do we need Winston (as maintainer) to kick off the revdepchecks / CRAN submission after that?

File: src/later.cpp
Comment (jcheng5) [reply src/later.cpp:213]: In that case, my very last question (hopefully)... `invoke_wrapped()` calls `invoked()`, which now includes `unwindProtect`/`BEGIN_CPP`. Does it make more sense to have `invoked_wrapped()` continue to call a code path that leaves that out? So maybe this conditional calls either `invoke_unwind_protect()` or `invoke_wrapped()`, each of which then call `invoke()`? Or is it OK the way it is?

File: src/later.cpp
Comment (shikokuchuo) [reply src/later.cpp:210]: Oh that's interesting. Because C++ code would always include `later_api.h`. Practically though I don't think it makes much difference to just asking users to use BEGIN_RCPP/END_RCPP. If we could define a class that does that with RAII it would save one line, but it'd still need to be instantiated within each function the user writes.

It's a cognitive burden either way. My thoughts were that if it hasn't surfaced for the time `later` has been in existence then it's probably not a major concern?

File: src/later.cpp
Comment (shikokuchuo) [reply src/later.cpp:213]: It's fine the way it is as the 2 code paths are completely separate. `invoke_wrapped()` is only called when unwind protect is not defined, in which case the invoke method itself will not be wrapped in `unwindProtect`.

With unwind protect: callback->invoke | invoke including `unwindProtect/BEGIN_CPP`
No unwind protect: callback->invoke_wrapped | invoke_c | invoke (without `unwindProtect`)

**Comment by shikokuchuo** [comment]
As a result of testing on a selection of rhub platforms, have made a couple of further changes to get this into shape:
1. [d0c8c0c](https://github.com/r-lib/later/pull/192/commits/d0c8c0c98a20ecc6ce240806be4c71aca5e17862) fixes compiler warning seen on rchk machine [here](https://github.com/shikokuchuo/later/actions/runs/11963400571/job/33353774716#step:6:42)
2. [cfa562b](https://github.com/r-lib/later/pull/192/commits/cfa562bd6252a8a75af862d081d337fc2c1e94db) skips the final block of C++ error and exception handling on CRAN as this was always skipped before and adding it in is just asking for trouble.
We gain nothing from doing so, and we still test on CI.
I saw some strangeness on the last test on one of the platforms [here](https://github.com/shikokuchuo/later/actions/runs/11963400571/job/33353773093#step:6:256) where the interrupt was returning an error.
There is also a ubsan skip for this block, which is redundant, but just for clarity as there is also some noise [here](https://github.com/shikokuchuo/later/actions/runs/11963400571/job/33353773821#step:6:341).

**Review by shikokuchuo — APPROVED** [review]

**Review by shikokuchuo — APPROVED** [review]
I'm re-approving as I did make this commit [a16ac2c](https://github.com/r-lib/later/pull/192/commits/a16ac2c76772b358274d6a1953cb79e97e146903) to get rid of some noise in the valgrind output (it was still passing). It just runs loop finalizers on exit and results in no actual difference in behaviour.

**Comment by jcheng5** [comment]
Sorry for the delay, finally back home with power and internet restored.

**Comment by shikokuchuo** [comment]
Thanks! Hope everything is fine and you weren't otherwise affected.

**Pull request merged**
